Shared/Realm.Sync.Shared/User.cs

```diff
@@ -0,0 +1,152 @@
+using System.IO;
+using System.Json;
+using System.Net;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Runtime.Serialization;
+using System.Threading.Tasks;
+
+namespace Realms.Sync
+{
+    [Flags]
+    public enum LoginMode
+    {
+        UseExistingAccount = 1,
+        CreateAccount = 2
+    }
+
+    public enum UserState
+    {
+        LoggedOut,
+        Active,
+        Error
+    }
+
+    public class User
+    {
+        private static readonly MediaTypeHeaderValue _applicationJsonUtf8MediaType = MediaTypeHeaderValue.Parse("application/json; charset=utf-8");
+        private static readonly MediaTypeHeaderValue _applicationProblemJsonUtf8MediaType = MediaTypeHeaderValue.Parse("application/problem+json; charset=utf-8");
+
+        public string RefreshToken => Handle.RefreshToken;
+
+        public string Identity => Handle.Identity;
+
+        public Uri ServerUri
+        {
+            get
+            {
+                var serverUrl = Handle.ServerUrl;
+                if (string.IsNullOrEmpty(serverUrl))
+                {
+                    return null;
+                }
+
+                return new Uri(serverUrl);
+            }
+        }
+
+        public UserState State => Handle.State;
+
+        internal readonly SyncUserHandle Handle;
+
+        internal User(SyncUserHandle handle)
+        {
+            Handle = handle;
+        }
+
+        public static async Task<User> LoginAsync(Credentials credentials, Uri serverUrl, LoginMode loginMode = LoginMode.UseExistingAccount)
+        {
+            if (credentials.IdentityProvider == Credentials.Providers.AccessToken)
+            {
+                var identity = (string)credentials.UserInfo[Credentials.Keys.Identity];
+                var isAdmin = (bool)credentials.UserInfo[Credentials.Keys.IsAdmin];
+                return new User(SyncUserHandle.GetSyncUser(identity, credentials.Token, serverUrl?.AbsoluteUri, isAdmin));
+            }
+
+            var result = await MakeAuthRequestAsync(serverUrl, credentials.ToJson(), TimeSpan.FromSeconds(30)).ConfigureAwait(continueOnCapturedContext: false);
+            var refresh_token = result["refresh_token"];
+
+            return new User(SyncUserHandle.GetSyncUser(refresh_token["token_data"]["identity"], refresh_token["token"], serverUrl.AbsoluteUri, false));
+        }
+
+        public void LogOut()
+        {
+            Handle.LogOut();
+        }
+
+        // returns a tuple of access token and resolved realm path
+        internal async Task<Tuple<string, string>> RefreshAccessToken(string path)
+        {
+            var json = new JsonObject
+            {
+                ["data"] = RefreshToken,
+                ["path"] = path,
+                ["provider"] = "realm"
+            };
+
+            var result = await MakeAuthRequestAsync(ServerUri, json, TimeSpan.FromSeconds(30)).ConfigureAwait(continueOnCapturedContext: false);
+            var access_token = result["access_token"];
```

Comment:
You're right, but I find using the same name helps me reason about things better. We also use snake_case for identifiers that denote something from our C++ layer like `DllImport`-ed functions or marshaled structures, so I could argue there is precedent for this.